runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerWrapper.java

```diff
@@ -0,0 +1,127 @@
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/** A docker command wrapper. Simplifies communications with the Docker daemon. */
+class DockerWrapper {
+  // TODO: Should we require 64-character container ids? Docker technically allows abbreviated ids,
+  // but we _should_ always capture full ids.
+  private static final Pattern CONTAINER_ID_PATTERN = Pattern.compile("\\p{XDigit}{64}");
+
+  static DockerWrapper forCommand(String dockerExecutable, Duration commandTimeout) {
+    return new DockerWrapper(dockerExecutable, commandTimeout);
+  }
+
+  private final String dockerExecutable;
+  private final Duration commandTimeout;
+
+  private DockerWrapper(String dockerExecutable, Duration commandTimeout) {
+    this.dockerExecutable = dockerExecutable;
+    this.commandTimeout = commandTimeout;
+  }
+
+  /**
+   * Runs the given container image with the given command line arguments. Returns the running
+   * container id.
+   */
+  public String runImage(String imageTag, List<String> args)
+      throws IOException, TimeoutException, InterruptedException {
+    checkArgument(!imageTag.isEmpty(), "Docker image tag required");
+    // TODO: Validate args?
+    return runShortCommand(
+        ImmutableList.<String>builder()
+            .add(dockerExecutable)
+            .add("run")
+            .add("-d")
+            .add(imageTag)
+            .addAll(args)
+            .build());
+  }
+
+  /**
+   * Kills a docker container by container id.
+   *
+   * @throws IOException if an IOException occurs or if the given container id does not exist
+   */
+  public void killContainer(String containerId)
+      throws IOException, TimeoutException, InterruptedException {
+    checkArgument(containerId != null);
+    checkArgument(
+        CONTAINER_ID_PATTERN.matcher(containerId).matches(),
+        "Container ID must be a 64-character hexadecimal string");
+    runShortCommand(Arrays.asList(dockerExecutable, "kill", containerId));
+  }
+
+  /** Run the given command invocation and return stdout as a String. */
+  private String runShortCommand(List<String> invocation)
+      throws IOException, TimeoutException, InterruptedException {
+    ProcessBuilder pb = new ProcessBuilder(invocation);
+    Process process = pb.start();
+    // TODO: Consider supplying executor service here.
+    CompletableFuture<String> resultString =
+        CompletableFuture.supplyAsync(
+            () -> {
+              // NOTE: We do not own the underlying stream and do not close it.
+              BufferedReader reader =
+                  new BufferedReader(
+                      new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
```

Comment:
<!--thread_id:cc_182896313_t; commit:eb0cdf4ea2d094d8ab09a0a9ebd742507f71a1a8; resolved:0-->
<!--section:context-quote-->
> **jkff** wrote:
> Should we capture the error stream also?

<!--section:body-->
We could. I went through several iterations of this previously. We're only using stdout for container management, so it's not strictly necessary. I can add stderr output, e.g., to exceptions if that would be useful.